crypto/serializer/serializer_pkey.c

```diff
@@ -0,0 +1,361 @@
+ */
+
+struct serializer_write_data_st {
+    OSSL_SERIALIZER_CTX *ctx;
+    BIO *out;
+};
+
+static int serializer_write_cb(const OSSL_PARAM params[], void *arg)
+{
+    struct serializer_write_data_st *write_data = arg;
+    OSSL_SERIALIZER_CTX *ctx = write_data->ctx;
+    BIO *out = write_data->out;
+
+    return ctx->ser->serialize_data(ctx->serctx, params, out,
+                                    serializer_passphrase_out_cb, ctx);
+}
+
+/*
+ * Support for OSSL_SERIALIZER_to_bio:
+ * Perform the actual output.
+ */
+
+static int serializer_EVP_PKEY_to_bio(OSSL_SERIALIZER_CTX *ctx, BIO *out)
+{
+    const EVP_PKEY *pkey = ctx->object;
+    void *provdata = pkey->pkeys[0].provdata;
+    int domainparams = pkey->pkeys[0].domainparams;
+    EVP_KEYMGMT *keymgmt = pkey->pkeys[0].keymgmt;
+
+    /*
+     * Hidden trick: OSSL_SERIALIZER_CTX_new() creates a context, even
+     * when the serializer it's given is NULL.  That works as a signal
+     * for functions like this one that the operation is unsupported,
+     * and to return a value indicating that so the caller can fall back
+     * on other functionality, is such a thing is available.
+     */
+    if (ctx->ser == NULL)
+        return -2;
+
+    if (ctx->ser->serialize_object == NULL) {
+        struct serializer_write_data_st write_data;
+
+        write_data.ctx = ctx;
+        write_data.out = out;
+
+        if (domainparams)
+            return evp_keymgmt_exportdomparams(keymgmt, provdata,
+                                               serializer_write_cb,
+                                               &write_data);
+        return evp_keymgmt_exportkey(keymgmt, provdata,
+                                     serializer_write_cb, &write_data);
+    }
+
+    return ctx->ser->serialize_object(ctx->serctx, provdata, out,
+                                      serializer_passphrase_out_cb, ctx);
+}
+
+/*
+ * OSSL_SERIALIZER_CTX_new_by_EVP_PKEY() returns a ctx with no serializer if
+ * it couldn't find a suitable serializer.  This allows a caller to detect if
+ * a suitable serializer was found, with OSSL_SERIALIZER_CTX_get_serializer(),
+ * and to use fallback methods if the result is NULL.
+ */
+OSSL_SERIALIZER_CTX *OSSL_SERIALIZER_CTX_new_by_EVP_PKEY(const EVP_PKEY *pkey,
+                                                         const char *propquery)
+{
+    OSSL_SERIALIZER_CTX *ctx = NULL;
+    OSSL_SERIALIZER *ser = NULL;
+    EVP_KEYMGMT *keymgmt = pkey->pkeys[0].keymgmt;
+    int want_domparams = -1;
+    const char *s;
+
+    if (pkey == NULL || propquery == NULL) {
+        ERR_raise(ERR_LIB_OSSL_SERIALIZER, ERR_R_PASSED_NULL_PARAMETER);
+        return NULL;
+    }
+
+    if ((s = strstr(propquery, "type")) == NULL)
+        goto parse_end;
+    if (!(s == propquery || s[-1] == ',' || s[-1] == ' '))
+        goto parse_end;
+    for (s += 4; s[0] == ' '; s++)
+        continue;
+    if (s[0] != '=')
+        goto parse_end;
+    for (s += 4; s[0] == ' '; s++)
```

Comment:
+4 ??

This needs to be wrapped into the parser, there are too many missing cases.  For example:
```
type="domainparams"
type='domainparams'
? type = domainparams
```
are all legal (the latter is a bit nonsensical in context).

Is making a `type=domainparams` query against the fetched serialiser's properties a possibility?

Reply:
Argh, copy-n-paste from line 313...

The kind of check you're talking about has an interesting property...  it means that I'll have to take `pkey->pkeys[0].domainparams` on faith, but then use its value to limit what serializers are considered.  Yeah, sure, that can be done.

Thanks, I might even figure out on my own  how to do that (although, giving me a pointer at the function to call might be helpful).

Reply:
Still pass the original query through which should do the value limiting.  Just make a second query solely about the type to figure out what.

It means exposing the property parse/query internals a bit and keeping the property definitions associated with algorithms in an accessible manner.

Reply:
Ok, I've reworked this as you suggested.  It's not very ugly at all!

Reply:
Looks neat.